### ISSUE TITLE
How to get `torch.cuda` outputs from Yolo-NAS?

### ISSUE THREAD

**Issue by mikel-brostrom** [Deci-AI/super-gradients#1065]
### 💡 Your Question

No matter what I do the output is always of type: `numpy.ndarray`

```python
import torch
import numpy as np

from super_gradients.common.object_names import Models
from super_gradients.training import models


rgb = np.random.rand(640, 640, 3).astype(np.uint8)
rgb_torch = torch.from_numpy(rgb)

yolo_nas_model = models.get(Models.YOLO_NAS_S, pretrained_weights="coco").to("cuda")

prediction = next(iter(yolo_nas_model.predict(rgb, iou=0.5, conf=0.7))).prediction
print(type(prediction.bboxes_xyxy))
prediction = next(iter(yolo_nas_model.predict(rgb_torch, iou=0.5, conf=0.7))).prediction
print(type(prediction.bboxes_xyxy))
```

the output datatype for both inputs:

```bash
<class 'numpy.ndarray'>
<class 'numpy.ndarray'>
```

Is there a way to make to model output `torch.cuda` data when the model inferes on cuda?

### Versions

Versions of relevant libraries:
[pip3] numpy==1.23.0
[pip3] torch==1.13.1
[pip3] torchmetrics==0.8.0
[pip3] torchvision==0.14.1
[pip3] triton==2.0.0

**Comment by Louis-Dupont** [comment]
Hi @mikel-brostrom ,
You raised a good point. I wrote below an explanation of why we currently don't return torch tensors, let me know if you think otherwise :)

### Motivation behind using `np.ndarray`
We chose to use `np.ndarray` as the default representation for the images because it is very common for people to work directly with `opencv` images instead of `torch`. Eventually, this simplifies the flow of many users who don't need to think about torch, let the `model.predict()` method handle everything and just get their `np.ndarray` image ready for `opencv`. 


### Should we also support returning torch tensors ?
I am not sure if changing the output type based on input type would be great, I tend to think it is better to have a static output type because I definitely see people writing a code that would work for an `np.ndarray` images, then passing a `torch` tensor, getting an unexpected error and opening issues because of that. 


We could also add an extra parameter `model.predict(to='cuda')` for instance, but I feel like adding all this type of parameter to the predict would just overload it with task that are not really its core responsibility of the predict method. 
The post-processed outputs are in `np.ndarray` format anyway, so whether the user or the predict method moves it to `cuda` will have the exact same impact on the performance (which is small compared to `mode.predict` time anyway)

Eventually, I still tend to think that the best design is to:
- Stick to returning a fixed type (`np.ndarray`) for the reasons above
- Let the user have the responsibility to move to torch/cuda if wanted. `bboxes_xyxy=torch.from_numpy(np.array(prediction.bboxes_xyxy)).cuda()`

**Comment by mikel-brostrom** [comment]
Moving data from cpu to gpu is very expensive. Some users will most certainly want the output in `torch.cuda` format for further processing on GPU. If you do this automatically, you force these users to move data back to GPU with its corresponding extra computational cost.

**Comment by Louis-Dupont** [comment]
You are right, but I believe that in practice it is actually not an issue.

Let's run the following code
``` python
import time
import torch

from super_gradients.common.object_names import Models
from super_gradients.training import models

# LOAD MODEL
start_t = time.process_time()
model = models.get(Models.YOLO_NAS_L, pretrained_weights="coco")
model_loaded_time = time.process_time()
print(f"Model loaded in {model_loaded_time - start_t}")

# MODEL TO CUDA
model = model.cuda()
model_to_cuda_time = time.process_time()
print(f"Model to cuda in {model_to_cuda_time - model_loaded_time}")

image = "../../../../documentation/source/images/examples/countryside.jpg" # You can use another image, but ideally one with prediction, and not an url to not include downloading time which is not what we care about here

# Doing some warmup
for _ in range(20):
    _ = model.predict(image, iou=0.5, conf=0.7)

# PREDICT
start_predict_time = time.process_time()
prediction = next(iter(model.predict(image, iou=0.5, conf=0.7))).prediction
prediction_time = time.process_time()
print(f"Prediction in {prediction_time - start_predict_time}")

# PREDICT TO CUDA
bboxes_xyxy = torch.from_numpy(prediction.bboxes_xyxy).cuda()
prediction_to_cuda_time = time.process_time()
print(f"Prediction to cuda in {prediction_to_cuda_time - prediction_time}")
```

This is what I get:
```
Model loaded in 86.90403328
Model to cuda in 10.286816428000009
Prediction in 0.1976654329999974
Prediction to cuda in 0.0005071420000035687
```

What we can see is that:
1. Moving the model to cuda is definitely expensive. 
2. Moving the prediction to cuda is ~400 times faster than the predict itself, i.e. ~`0.25%` of the function call. This is because the prediction is an extremely small object compared to the model. This may vary on the number of predictions, but this should be enough for a rough estimation.

From there, we could argue that even if it is very small, it would still be optimized to not move it to `cuda` again, and it's true. But if a decrease in performance of even up to `1%` is a big deal, then the right way to go would be not to use the predict method, which only works on the torch `.pth` model, and to work with compiled model instead.

To summarize
- the current `predict` flow converts to `numpy` anyway, se don't have a simple workaround to avoid converting to `cuda`
- Yes, it is not ideal, but in practice, the performance impact is relatively small
- If the performance impact is too big for a use-case, then real win would not come from this but instead from using a compiled model

**Comment by harpreetsahota204** [comment]
Hi @mikel-brostrom 

Thanks for opening an issue for SG. I'm gathering some feedback on SuperGradients and YOLO-NAS.

Would you be down for a quick call to chat about your experience?

If a call doesn't work for you, no worries. I've got a short survey you could fill out: https://bit.ly/sgyn-feedback.

I know you’re super busy, but your input will help us shape the direction of SuperGradients and make it as useful as possible for you.

I appreciate your time and feedback. Let me know what works for you.

Cheers,

Harpreet

**Comment by mikel-brostrom** [comment]
Sure, I have some valuable feedback so let's chat :smile: ! Send me a meeting invite at [yolov5.deepsort.pytorch@gmail.com](mailto:yolov5.deepsort.pytorch@gmail.com)

**Comment by BloodAxe** [comment]
Closing this issue as no action is planned on the matter.
To summarie:

- predict() api is intended to take numpy images and return numpy boxes.
- predict() was never designed to be the fast inference option, rather an **easy** one.
- For best performance it is suggested to export models to ONNX and run in TRT

**Comment by natmurad** [comment]
Were you able to convert the output to tensors and not using predict?

I am having a hard time trying to convert the output and use GradCAM on YOLO-NAS model.

https://jacobgil.github.io/pytorch-gradcam-book/Class%20Activation%20Maps%20for%20Object%20Detection%20With%20Faster%20RCNN.html

**Comment by BloodAxe** [comment]
I suggest to go over documentation on the output predictions format for YoloNAS: https://docs.deci.ai/super-gradients/latest/documentation/source/ObjectDetection.html#understanding-models-predictions

The predictions of YoloNAS are result of concatenation and permutation of predictions from individual feature maps so you would need to undo this step from GradCAM.

In YoloNAS we have three feature maps [B,C,H//8,W//8], [B,C,H//16,W//16], [B,C,H//32,W//32]
This is what happens to them:
- [B,C,H//8,W//8] -> [B,C,A1] -> [B,A1,C]
- [B,C,H//16,W//16] -> [B,C,A2] -> [B,A2,C]
- [B,C,H//32,W//32] -> [B,C,A3] -> [B,A3,C]
Last step is concatenation: [B,A1,C], [B,A2,C], [B,A3,C] -> [B,A1+A2+A3,C]
So if you want to have gradcam you probably have to undo these steps or modify a heads. Whatever would be easier for you
